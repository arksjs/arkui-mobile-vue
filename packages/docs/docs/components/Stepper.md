# Stepper 步进器

注：

- 支持表单，具体可参考 [Form](./Form.md)。

<CodeDemo name="Stepper">

<<< @/../../demo/src/components/Form/Stepper/index.vue

</CodeDemo>

## Import

```js
import { Stepper } from 'vfox'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

## Props

| 属性           | 类型    | 默认值   | 必填 | 说明                                     |
| -------------- | ------- | -------- | ---- | ---------------------------------------- |
| v-model        | number  |          | 否   |
| name           | string  |          | 否   | 标识                                     |
| min            | number  | 0        | 否   | 最小值                                   |
| max            | number  | Infinity | 否   | 最大值                                   |
| step           | number  | 1        | 否   | 步长                                     |
| disabled       | boolean | false    | 否   | 是否禁用                                 |
| disabled-plus  | boolean | false    | 否   | 是否禁用增加按钮                         |
| disabled-minus | boolean | false    | 否   | 是否禁用减少按钮                         |
| disabled-input | boolean | false    | 否   | 是否禁用输入框输入                       |
| show-value     | boolean | false    | 否   | 是否显示当前 value                       |
| allow-decimal  | boolean | true     | 否   | 是否允许输入小数，`false` 限制只允许整数 |

## Events

| 事件        | 描述               | 回调函数参数            |
| ----------- | ------------------ | ----------------------- |
| change      | 当绑定值变化时触发 | value: number \| string |
| input       | 输入框输入时触发   | value: number \| string |
| focus       | 输入框聚焦时触发   | e: FocusEvent           |
| blur        | 输入框失焦时触发   | e: FocusEvent           |
| plus-click  | 点击增加按钮时触发 |                         |
| minus-click | 点击减少按钮时触发 |                         |
