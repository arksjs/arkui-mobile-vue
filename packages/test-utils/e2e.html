<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
