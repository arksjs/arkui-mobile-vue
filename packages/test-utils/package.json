{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "UI test-utils",
  "main": "dist/index.js",
  "private": true
}
