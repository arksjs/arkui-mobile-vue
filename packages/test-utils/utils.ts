export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
