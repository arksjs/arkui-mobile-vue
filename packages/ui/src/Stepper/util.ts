import {
  formatInputNumber,
  getNumber,
  rangeInteger,
  rangeNumber
} from '../helpers'

export const getClasses = (disabled?: boolean) => {
  return ['ta-stepper', { disabled: !!disabled }]
}

export function formatNumber(
  value: string | number,
  decimalLength?: number | string
): string {
  return formatInputNumber(value, getNumber(decimalLength, -1))
}

export function getRangeNumber(
  props: {
    min: number
    max: number
    decimalLength?: number | string
  },
  value: string | number
) {
  value = formatNumber(value, props.decimalLength)

  if (value === '') {
    value = props.min
  }

  if (props.decimalLength != 0) {
    value = rangeNumber(parseFloat(value as string), props.min, props.max)
  } else {
    value = rangeInteger(Math.floor(value as number), props.min, props.max)
  }

  return value.toString()
}
