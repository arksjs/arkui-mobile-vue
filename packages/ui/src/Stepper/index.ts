import { withInstall } from '../helpers/with-install'
import Stepper from './Stepper.vue'

export { Stepper }
export const install = withInstall(Stepper)
export default Stepper
