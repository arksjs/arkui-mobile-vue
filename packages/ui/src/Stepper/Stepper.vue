<template>
  <div :class="classes">
    <Button
      :icon="MinusOutlined"
      shape="square"
      size="small"
      :disabled="disabled || disabledMinus || parseFloat(formValue) <= nMin"
      @click="onMinusClick"
    ></Button>
    <input
      class="ak-stepper_input"
      :type="allowDecimal ? 'text' : 'tel'"
      :inputmode="allowDecimal ? 'decimal' : 'numeric'"
      :name="name"
      :disabled="disabled"
      :readonly="disabledInput"
      @input="onInput"
      @focus="onFocus"
      @blur="onBlur"
      @change="onChange"
      ref="inputEl"
    />
    <Button
      :icon="PlusOutlined"
      shape="square"
      size="small"
      :disabled="disabled || disabledPlus || parseFloat(formValue) >= nMax"
      @click="onPlusClick"
    ></Button>
  </div>
</template>

<script lang="ts">
import { onMounted, ref, defineComponent, watch, computed } from 'vue'
import { Button } from '../Button'
import { getNumber } from '../helpers/util'
import { formateNumber, getRangeNumber, getClasses } from './util'
import {
  formStringValueEmits,
  formItemProps,
  formFocusEmits
} from '../Form/form'
import { useInput } from '../Form/use-form'
import PlusOutlined from '../Icon/icons/PlusOutlined'
import MinusOutlined from '../Icon/icons/MinusOutlined'
import type { OnFocus, OnClick, PropsToEmits } from '../helpers/types'
import { emitClickValidator } from '../helpers/validator'
import { StepperEmits } from './types'

export default defineComponent({
  name: 'ak-stepper',
  components: { Button },
  props: {
    ...formItemProps,
    modelValue: {
      type: [Number, String]
    },
    // 是否禁用减少按钮
    disabledMinus: {
      type: Boolean,
      default: false
    },
    // 是否禁用增加按钮
    disabledPlus: {
      type: Boolean,
      default: false
    },
    // 是否禁用输入框
    disabledInput: {
      type: Boolean,
      default: false
    },
    min: {
      type: [Number, String],
      default: 1
    },
    max: {
      type: [Number, String],
      default: Infinity
    },
    allowDecimal: {
      type: Boolean,
      default: true
    },
    step: {
      type: [Number, String],
      default: 1
    },
    decimalLength: {
      type: Number
    }
  },
  emits: {
    ...formStringValueEmits,
    ...formFocusEmits,
    plusClick: emitClickValidator,
    minusClick: emitClickValidator
  } as PropsToEmits<StepperEmits>,
  setup(props, { emit }) {
    const formValue = ref('')
    const nMin = computed(() => getNumber(props.min, 1))
    const nMax = computed(() => getNumber(props.max, Infinity))
    const nStep = computed(() => getNumber(props.step, 1))

    const { inputEl, setInputValue, getInputValue } = useInput()

    const classes = computed(() => getClasses(props.disabled))

    function updateValue(val: number | string) {
      const newVal = getRangeNumber(
        {
          min: nMin.value,
          max: nMax.value,
          allowDecimal: props.allowDecimal,
          decimalLength: props.decimalLength
        },
        val
      )

      if (newVal != props.modelValue) {
        emit('update:modelValue', newVal)
      }

      if (newVal !== formValue.value) {
        formValue.value = newVal

        emit('change', newVal)
      }

      setInputValue(newVal)

      return newVal
    }

    const onFocus: OnFocus = e => emit('focus', e)
    const onBlur: OnFocus = e => emit('blur', e)

    function onChange() {
      updateValue(getInputValue())
    }

    function onInput() {
      const val = formateNumber(getInputValue())
      setInputValue(val)

      emit('input', val)
    }

    const onMinusClick: OnClick = e => {
      updateValue(parseFloat(formValue.value) - nStep.value)

      emit('minusClick', e)
    }

    const onPlusClick: OnClick = e => {
      updateValue(parseFloat(formValue.value) + nStep.value)

      emit('plusClick', e)
    }

    onMounted(() => {
      // 同步Mounted之前设置的值到 input 元素
      setInputValue(formValue.value)
    })

    watch(
      () => props.modelValue,
      val => {
        if (
          val != null &&
          parseFloat(val.toString()) !== parseFloat(formValue.value)
        ) {
          updateValue(val)
        } else if (formValue.value === '') {
          // 针对首次没有值的情况，默认最小值
          updateValue(nMin.value)
        }
      },
      {
        immediate: true
      }
    )

    return {
      classes,
      inputEl,
      formValue,
      onFocus,
      onBlur,
      onChange,
      onInput,
      onMinusClick,
      onPlusClick,
      PlusOutlined,
      MinusOutlined,
      nMin,
      nMax
    }
  }
})
</script>
