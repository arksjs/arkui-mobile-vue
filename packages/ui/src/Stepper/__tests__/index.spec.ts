import { describe, test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { Stepper } from '@/Stepper'

describe('Stepper', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Stepper, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('disabled=true', () => {
    const wrapper = mount(Stepper, {
      props: {
        disabled: true
      }
    })

    expect(wrapper.classes()).toContain('disabled')
    expect(wrapper.find('.ak-stepper_input').attributes('disabled')).toBe('')

    const $buttons = wrapper.findAllComponents('.ak-button')

    expect($buttons[0].attributes('disabled')).toBe('')
    expect($buttons[1].attributes('disabled')).toBe('')
  })

  test.concurrent('disabledPlus & disabledMinus', async () => {
    const wrapper = mount(Stepper, {
      props: {
        disabledPlus: true,
        max: 5
      }
    })

    const $buttons = wrapper.findAllComponents('.ak-button')

    expect($buttons[0].attributes('disabled')).toBe('') // 当前处于最小值
    expect($buttons[1].attributes('disabled')).toBe('')

    await wrapper.setProps({
      modelValue: '2'
    })

    expect($buttons[0].attributes('disabled')).toBeUndefined()
    expect($buttons[1].attributes('disabled')).toBe('')

    await wrapper.setProps({
      disabledMinus: true,
      disabledPlus: false
    })

    expect($buttons[0].attributes('disabled')).toBe('')
    expect($buttons[1].attributes('disabled')).toBeUndefined()

    await wrapper.setProps({
      modelValue: '10'
    })

    expect($buttons[0].attributes('disabled')).toBe('')
    expect($buttons[1].attributes('disabled')).toBe('') // 当前处于最大值
  })

  test.concurrent('min max', async () => {
    const wrapper = mount(Stepper, {
      props: {
        min: 1,
        max: 5
      }
    })

    expect(wrapper.vm.formValue).toBe('1')

    // 按钮操作
    const $buttons = wrapper.findAllComponents('.ak-button')

    await $buttons[0].trigger('click')
    expect(wrapper.vm.formValue).toBe('1')

    await $buttons[1].trigger('click')
    expect(wrapper.vm.formValue).toBe('2')

    await $buttons[1].trigger('click')
    await $buttons[1].trigger('click')
    await $buttons[1].trigger('click')
    expect(wrapper.vm.formValue).toBe('5')

    await $buttons[1].trigger('click')
    expect(wrapper.vm.formValue).toBe('5')

    // 输入框操作
    await wrapper.find('input').setValue('3')
    expect(wrapper.vm.formValue).toBe('3')

    await wrapper.find('input').setValue('10')
    expect(wrapper.vm.formValue).toBe('5')
  })

  test.concurrent('step=0.1', async () => {
    const wrapper = mount(Stepper, {
      props: {
        min: 1,
        max: 5,
        step: 0.1
      }
    })

    expect(wrapper.vm.formValue).toBe('1')

    // 按钮操作
    const $buttons = wrapper.findAllComponents('.ak-button')

    await $buttons[1].trigger('click')
    expect(wrapper.vm.formValue).toBe('1.1')
  })

  test.concurrent('decimalLength=1', async () => {
    const wrapper = mount(Stepper, {
      props: {
        decimalLength: 1
      }
    })

    expect(wrapper.vm.formValue).toBe('1')

    await wrapper.setProps({
      modelValue: '1.1'
    })
    expect(wrapper.vm.formValue).toBe('1.1')

    await wrapper.setProps({
      modelValue: '1.11'
    })
    expect(wrapper.vm.formValue).toBe('1.1')
  })
})
