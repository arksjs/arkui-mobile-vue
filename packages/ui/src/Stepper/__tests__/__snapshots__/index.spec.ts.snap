// Vitest Snapshot v1

exports[`Stepper > snapshot 1`] = `
"<div class=\\"ta-stepper\\">
  <ta-button-stub icon=\\"[object Object]\\" shape=\\"square\\" size=\\"small\\" loading=\\"false\\" disabled=\\"true\\" ghost=\\"false\\" transparent=\\"false\\"></ta-button-stub><input class=\\"ta-stepper_input\\" type=\\"text\\" inputmode=\\"decimal\\" name=\\"\\">
  <ta-button-stub icon=\\"[object Object]\\" shape=\\"square\\" size=\\"small\\" loading=\\"false\\" disabled=\\"false\\" ghost=\\"false\\" transparent=\\"false\\"></ta-button-stub>
</div>"
`;
